source/common/router/config_impl.cc

```diff
@@ -523,14 +539,18 @@ const VirtualHostImpl* RouteMatcher::findVirtualHost(const Http::HeaderMap& head
     return default_virtual_host_.get();
   }
 
-  auto iter = virtual_hosts_.find(headers.Host()->value().c_str());
+  const char* host = headers.Host()->value().c_str();
+  auto iter = virtual_hosts_.find(host);
```

Comment:
headers.Host()->value() actually returns a HeaderString which just happens to have a c_str method.